### PR TITLE
SOR - Add support for paths with buffers/boosted pools

### LINKED ISSUE #557: V3 SOR - Add buffer/boosted pool support
SOR needs to consider buffer/boosted pools and return appropriate path data (will be handled by SDK team).

I'm [adding SDK support](https://github.com/balancer/b-sdk/pull/348) assuming:
SOR will add additional field: `isBuffer: boolean[]` as part of `Path` return. Each isBuffer array item at index, `i`, should indicate if corresponding pool in `pools[i]` is a buffer pool. If not returned/undefined we assume there are no buffers in swap path.

An example `Path`:
```
Path = {
    protocolVersion: 3,
    tokens: [
        {
            address:
                '0x94a9D9AC8a22534E3FaCa9F4e7F2E2cf85d5E4C8',
            decimals: 18,
        }, // DAI
        {
            address:
                '0x8a88124522dbbf1e56352ba3de1d9f78c143751e',
            decimals: 18,
        }, // stataToken
        {
            address:
                '0xde46e43f46ff74a23a65ebb0580cbe3dfe684a17',
            decimals: 6,
        }, // stataToken
        {
            address:
                '0xFF34B3d4Aee8ddCd6F9AFFFB6Fe49bD371b8a357',
            decimals: 6,
        }, // USDC
    ],
    pools: [
        '0x8a88124522dbbf1e56352ba3de1d9f78c143751e', // buffer
        '0x90a46864cb1f042060554592038367e9c97e17f3',
        '0xde46e43f46ff74a23a65ebb0580cbe3dfe684a17', // buffer
    ],
    isBuffer: [true, false, true],
    inputAmountRaw: 1000000000000000000n,
    outputAmountRaw: 1000000n,
}
```

### PR DESCRIPTION
**Pull request by brunoguerios**
Closes #557 

Here's a brief explanation on the approach implemented on this PR to update the SOR in order to support paths with boosted/buffer pools.

Basically it includes a new BufferPool for each existing unique erc4626 token, which will behave as if it's facilitating a trade between a yield bearing token and its underlying token (e.g. stataDAI/DAI).

With that in place, the SOR is now able to create swap paths with wrap/unwrap steps without any extra change.

Notes:
- It's possible that under the hood there's no real buffer with liquidity to facilitate those trades, but in that scenario the Vault is able to reach out to the external protocol to perform a wrap/unwrap operation. The SOR is unaware of that and this does not affect the SOR result.
- All pool math complexity is handled within Balancer Maths and data fetching is already handled by the API.

### CODE REVIEW

**Comment by brunoguerios** [comment]
Checks seem to be failing due to an unrelated issue. Will check again once things are fixed on v3-canary

File: modules/sor/balancer-sor.integration.test.ts
Comment (johngrantuk) [inline modules/sor/balancer-sor.integration.test.ts:1]: Seeing some warning for `sorGetPathsWithPools` in places, probably worth fixing.

File: modules/sor/balancer-sor.integration.test.ts [hunk 1]
```diff
@@ -329,6 +329,79 @@ describe('Balancer SOR Integration Tests', () => {
...
 
+    describe('Buffer Pool Path', () => {
+        beforeAll(async () => {
+            // setup mock pool data
+            const poolAddress = '0x302b75a27e5e157f93c679dd7a25fdfcdbc1473c';
+            const stataUSDC = prismaPoolTokenFactory.build({
+                address: '0x8a88124522dbbf1e56352ba3de1d9f78c143751e',
+                token: { decimals: 6, underlyingTokenAddress: '0x94a9d9ac8a22534e3faca9f4e7f2e2cf85d5e4c8' },
+                dynamicData: prismaPoolTokenDynamicDataFactory.build({
+                    balance: '500',
```
Comment (johngrantuk) [inline modules/sor/balancer-sor.integration.test.ts:379]: Do you use Tenderly to get the pool state? Might be worth adding a bit of documentation at the top of the file explaining the process so its easier for the next person?

File: modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
+import { Address, Hex } from 'viem';
+import { MAX_UINT256, SwapKind, Token, TokenAmount } from '@balancer/sdk';
+import { BufferState, Vault } from '@balancer-labs/balancer-maths';
+
+import { BasePoolV3 } from '../../poolsV2/basePool';
+import { BasePoolToken } from '../../poolsV2/basePoolToken';
+import { Erc4626PoolToken } from '../../poolsV2/erc4626PoolToken';
+
```
Comment (johngrantuk) [inline modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts:8]: Would be helpful to add a bit of documentation here explaining what how the BufferPool and ERC4626 relate.

File: modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        return new BufferPool(
+            erc4626Token.token.address,
+            erc4626Token.token.address,
+            erc4626Token.token.chainId,
+            erc4626Token.rate,
+            [mainToken, underlyingToken],
+        );
+    }
+
+    constructor(id: Hex, address: string, chainId: number, rate: bigint, tokens: BasePoolToken[]) {
```
Comment (johngrantuk) [inline modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts]: This should only ever be constructed with a `mainToken` and an `underlyingToken` right? So we could explicitly use those as constructor params instead of `token[]` to make it more clear?

File: modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            poolType: 'Buffer',
+            poolAddress: this.address,
+            tokens: this.tokens.map((t) => t.token.address),
+            rate: this.rate,
+        };
+    }
+
+    // Helper methods
+
+    public getRequiredTokenPair(tokenIn: Token, tokenOut: Token): { tIn: BasePoolToken; tOut: BasePoolToken } {
```
Comment (johngrantuk) [inline modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts]: For some reason I find it hard to reason about whats happening here - could you explain and maybe refactor/rename to make it a bit clearer?

File: modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            },
+            poolState,
+        );
+        return TokenAmount.fromRawAmount(tOut.token, calculatedAmount);
+    }
+
+    public swapGivenOut(tokenIn: Token, tokenOut: Token, swapAmount: TokenAmount): TokenAmount {
+        const { tIn, tOut } = this.getRequiredTokenPair(tokenIn, tokenOut);
+
+        const poolState = this.getPoolState();
```
Comment (johngrantuk) [inline modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts]: `poolState` and `vault` never change after initial instantiation do they? So we could maybe just call them once in constructor and reuse?

**Review by johngrantuk — CHANGES_REQUESTED** [review]
Can you add an ELI5 description about the approach to the PR description 🙏
(@mkflow27 - adding you as a reviewer as it will help you understand the SOR stuff a bit more I think)

File: modules/sor/balancer-sor.integration.test.ts
Comment (brunoguerios) [reply modules/sor/balancer-sor.integration.test.ts:1]: Fixed ✅

File: modules/sor/balancer-sor.integration.test.ts (hunk 1 above)
Comment (brunoguerios) [reply modules/sor/balancer-sor.integration.test.ts:379]: Yeah, this was a pretty manual process. I'll add a bit of documentation to get this PR ready, but I'll open another issue aiming to streamline this testData process. Ideally with something similar to Balancer Maths repo.

File: modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts (hunk 4 above)
Comment (brunoguerios) [reply modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts]: This is actually doing a two things:
1. validating that the pool indeed contains tokenIn and tokenOut
2. getting pool tokens with specific attributes (e.g. weight for WeightedPool, rate for StablePool) and their state (balances, index, etc..) for tokenIn and tokenOut

I agree the name could be more clear on the function's purpose.
Do you have any specific naming suggestions?
I couldn't think of a better naming for it, but I'll add some comments explaining what it's doing.
Another thing that I think might help is to add it to the BasePool interface to highlight that this is present on all pools.

Let me know what you think.

**Comment by brunoguerios** [comment]
I applied most suggestions and left a few responses on the comments. 
It's still missing the PR ELI5 description and some testData docs.
I'll work on them tomorrow morning.

File: modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts (hunk 4 above)
Comment (johngrantuk) [reply modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts]: Ahh right, got it, thanks! Comments will definitely help clarify.

Maybe something like `getBasePoolTokens` might make it clearer?

Looks like its just using the wrappedToken address so might also make sense to make the input params `0x${string}` types instead of `Token` and rename to something like wrappedTokenIn/Out? Can you remind me of what the `wrapped` param represents and what it would be for something like stataUSDC? Thanks!

Comment (brunoguerios) [reply modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts]: I think I'll go with `getPoolTokens` so it's more comprehensive for all PoolToken types.

About updating the inputs, I don't think there's much benefit from it. This method is used only internally on each pool, so I think it's simpler to stick with forwarding Token from external inputs (e.g. swapGivenIn, swapGivenOut, etc.). I can revert it back to private and remove it from the interface if you think it makes more sense.

`wrapped` is used only in the context of native assets and their wrapped versions (e.g. ETH/WETH). For all other tokens, `address` and `wrapped` are equal.

File: modules/sor/balancer-sor.integration.test.ts (hunk 1 above)
Comment (brunoguerios) [reply modules/sor/balancer-sor.integration.test.ts:379]: Done ✅ 
I hope to find some time to improve this on the next few weeks. It's quite annoying to update manually all that test data.
On top of that, subgraph prunes old blocks, so we're not even able to verify the data if needed 😕

File: modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts (hunk 4 above)
Comment (johngrantuk) [reply modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts]: Cool. Just one last question on this - doesn't look like any of the other pools are checking `wrapped` should they?

**Review by johngrantuk — APPROVED** [review]
LGTM 🙏

File: modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts (hunk 4 above)
Comment (brunoguerios) [reply modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts]: Yes, you're right. We should be checking wrapped everywhere. Very few places should check for address directly. I'll fix that over here.

Comment (brunoguerios) [reply modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts]: Although, looking again at the code, it seems the API is handling this differently and replacing ETH with WETH inputs so this `wrapped` concept may not even make much sense anymore in the current scenario. I'll create a separate task to investigate this and possibly remove that from SDK/SOR Token type.
